---
title: "INFO 550: Project"
author: "Randy Parrish"
date: "`r format(Sys.Date())`"
output: 
  html_document:
    pandoc_args: [
      "--bibliography", "biblio.bib",
      "--csl", "citestyle.csl",
      "--metadata", "link-citations=true"
      ]
params: 
  usebiblio: 1
  usecache: 0
---

```{r}
# print(metadata$output$html_document$pandoc_args)
# print(as.list(.GlobalEnv))
print(rmarkdown::output_format)
```

<!-- Load libraries, set up chunk options -->
```{r, setup, include = FALSE}
Sys.setlocale('LC_ALL', 'C')

library(ggplot2)
library(knitr)
library(kableExtra)

opts_chunk$set(eval = TRUE, echo = FALSE, warning = FALSE, 
    dpi = 300, fig.align = 'left', cache = params$usecache)
```

<!-- Define functions -->
```{r, define-func}

# generates qq plot of gene p-values
make_qq_plot <- function(pvector, size=8) {
    pvector = pvector[!is.na(pvector)]
    o = -log10(sort(pvector, decreasing=F))
    e = -log10( 1:length(o)/length(o) )
    max_lim <- max(c(max(e), max(o[o<Inf])))
    qplot(e, o) + xlim(c(0, max_lim)) + ylim(c(0, max_lim)) + 
        labs(x = "Expected", y = "Observed") + 
        geom_abline(intercept = 0, slope = 1, colour = "red") + 
    theme(text = element_text(size = size))
}

# generates manhattan plot of gene log p-values, line drawn at genome-wide significance level
make_man_plot <- function(manPlot_dt, chr_vec = 1:22, ntop = 3, 
		sig_level = 2.5e-6, size=10, chrGAP = 500, 
		colors=c('#44B5AD', '#3A948E', '#36807A', '#2f615d')){

    sig_label = paste0('-log(',sig_level,')')
    endPos = 0;
    plotPos = NULL; temp_dt = NULL; 
    chrEnd = NULL; LabBreaks = NULL; xlabels = NULL;
    for (chr in chr_vec) {
        temp = manPlot_dt[manPlot_dt$CHR == chr, ]
        temp$POS = order(temp$POS)
        if(nrow(temp)>0){
            temp_dt = rbind(temp_dt, temp)  
            chrPos = (temp$POS - min(temp$POS, na.rm = TRUE) ) + endPos + 1
            endPos = max(chrPos, na.rm = TRUE) + chrGAP
            plotPos = c(plotPos, chrPos)
            yline_pos = max(chrPos, na.rm = TRUE) + chrGAP/2
            chrEnd = c(chrEnd, yline_pos)
            LabBreaks = c(LabBreaks, mean(chrPos, na.rm = TRUE) )
            xlabels = c(xlabels, chr)
        }
    }
    manPlot_dt_sort = data.frame(plotPos = plotPos, temp_dt)
    ggplot(manPlot_dt_sort, 
    	aes(plotPos, -log10(PVALUE), color=factor(CHR), label=ID)) + 
    geom_point(size=2,alpha=0.9) + 
    guides(color = FALSE) + 
    scale_color_manual(values = rep(colors, 22)) +
    geom_hline(yintercept = -log10(sig_level), 
    	linetype="dashed", size=1, color = "black") +
    labs(x = "Chromosome", y = "-log10(PVALUE)") + 
    scale_x_continuous(breaks = LabBreaks, 
    	labels = xlabels, expand=c(0.01, 0)) + 
    coord_cartesian(clip = 'off') +
    scale_y_continuous(expand = c(0.01, 0)) +
    theme(text = element_text(size = size), 
        axis.text.x = element_text(face = "bold", 
        	size=size-1, angle=-90, vjust=0.5, hjust=0),
        panel.grid.major.x = element_blank(),
        panel.grid.minor.x = element_blank(),
        panel.border = element_blank()
        )
}

# generate table of significant genes from provided data
make_sig_table <- function(data){
	return(scroll_box(
		kable_paper(
			kable_styling(
				kbl(data, row.names = FALSE), 
				full_width = F, position = 'center',
				bootstrap_options = c('striped','hover')), 
			full_width = F), 
		fixed_thead = T, width = '400px', height = '500px'))
}

# globally sets dataframes used to generate figures/plots for given TWAS 
get_twas_data <- function(twas_name){
	twas_data <- data[data$TWAS == twas_name, c(2:5,7)]

	# data for plots
	plot_data <<- twas_data[, c(1,2,4,5)]
	colnames(plot_data) <<- c('CHR','POS','ID','PVALUE')

	# data for significant genes table
	sig_data <<- twas_data[twas_data$PVALUE <= 2.5e-6, c(4,1:3,5)]
	colnames(sig_data) <<- c('Gene','Chrom','Start','End','Pvalue')
	sig_data['Pvalue'] <<- format(sig_data[['Pvalue']], scientific = TRUE, digits = 3)
	twas <<- twas_name
}
```

<!-- Load data -->
```{r, load-data}
# data file should be in same directory as Rmd file
data <- read.table('./TWAS_data.txt', 
	header = TRUE, sep = '\t', row.names = 1, skipNul = TRUE,
	col.names = c('TWAS','CHROM','GeneStart',
		'GeneEnd', 'TargetID', 'GeneName',
		'n_snps','ZScore', 'PVALUE'),
	colClasses = c('integer','character', 'integer', 
		'integer', 'integer', 'NULL', 
		'character', 'NULL', 'numeric', 'NULL'))

data['PVALUE'] <- exp(pchisq(data[['ZScore']]^2, 1, lower.tail=FALSE, log.p = TRUE))

# sort by TWAS, pvalue, chrom
data <- data[order(data$TWAS, data$PVALUE, data$CHROM), ]
```

# Background

The software tool TIGAR (Transcriptome-Integrated Genetic Association Resource) [@nagpalTIGARImprovedBayesian2019] was used to train gene imputation models using the DPR model and conduct summary-level association studies using a burden Z-score statistic as implemented by S-PrediXcan [@barbeiraExploringPhenotypicConsequences2018].

Genotype and gene expression data from version 8 of the Genotype-Tissue Expression (GTEx) project [@carithersNovelApproachHighQuality2015a] were used to train gene imputation models for 140 normal ovary tissue samples and 337 normal breast tissue samples from subjects of European ancestry.

Gene expression data was adjusted for eQTL analysis covariates, including principal components and PEER factors, provided by GTEx as well as age and BMI. Breast mammary tissue expression data was also adjusted for *ESR1* expression. 

Models were trained for genes with expression thresholds of $>0.1$ TPM in $\geq10$ samples using variants with minor allele frequency $>0.01$ and  Hardy-Weinberg equilibrium p-value $>0.00001$. Only gene models with an average 5-fold cross-validation $R^2>0.01$ were retained.

Breast and Ovarian cancer GWAS Z-score summary-statistics were obtained from the Breast Cancer Association Consortium (BCAC) [@michailidouAssociationAnalysisIdentifies2017] and the Ovarian Cancer Association Consortium (OCAC) [@phelanIdentificationTwelveNew2017], respectively.


# Results

<!-- Breast Cancer, Breast Tissue TWAS Results -->
## Breast Cancer TWAS

```{r, breast-data}
get_twas_data('Breast')
```

### Manhattan Plot

```{r, breast-twas-man, fig.width = 5, fig.asp = 0.6}
make_man_plot(plot_data)
```

### QQ Plot

```{r, breast-twas-qq, fig.width = 1.75, fig.asp = 1}
make_qq_plot(plot_data[['PVALUE']])
```

### Significant Genes

```{r, breast-twas-table}
make_sig_table(sig_data)
```

<!-- Ovarian Cancer, Ovary Tissue TWAS Results -->
## Ovarian Cancer TWAS

```{r, ovary-data}
get_twas_data('Ovary')
```

### Manhattan Plot

```{r, ovary-twas-man, fig.width = 5, fig.asp = 0.6}
make_man_plot(plot_data)
```

### QQ Plot

```{r, ovary-twas-qq, fig.width = 1.75,  fig.asp = 1}
make_qq_plot(plot_data[['PVALUE']])
```

### Significant Genes

```{r, ovary-twas-table}
make_sig_table(sig_data)
```


<!-- Bibliography automatically generated from bib, csl files -->
<!-- Chunk used to avoid printing header if bibliography not created -->
```{r, bibliography-header, eval = params$usebiblio, results = 'asis', cache = FALSE}
cat('## Bibliography\\n')
```

